File: server/models/move.py
from typing import List
from pydantic import BaseModel

from .constants import Piece

class Movement(BaseModel):
    piece: Piece
    srcx: int
    srcy: int
    dstx: int
    dsty: int
    take: Piece

    @classmethod
    def parse(self, data):
        return Movement(
            piece = data['piece'],
            srcx = data['srcx'],
            srcy = data['srcy'],
            dstx = data['dstx'],
            dsty = data['dsty'],
            take = data['take'],
        )


class Move(BaseModel):
    movements: List[Movement]

    @classmethod
    def parse(self, data):
        return Move(movements = [Movement.parse(mv) for mv in data['movements']])

File: client/src/models/move.ts
import {Piece} from './constants';

export class Movement {
    public piece: Piece;
    public srcx: number;
    public srcy: number;
    public dstx: number;
    public dsty: number;
    public take: Piece;

    public static parse(json) {
        return new Movement(json.piece, json.srcx, json.srcy, json.dstx, json.dsty, json.take);
    }

    constructor(piece?: Piece, srcx?: number, srcy?: number, dstx?: number, dsty?: number, take?: Piece) {
        this.piece = piece;
        this.srcx = srcx;
        this.srcy = srcy;
        this.dstx = dstx;
        this.dsty = dsty;
        this.take = take;
    }

    public equals(other: Movement) {
        if (!other) {
            return false;
        }

        return this.piece === other.piece && this.srcx === other.srcx && this.srcy === other.srcy && this.dstx === other.dstx && this.dsty === other.dsty && this.take === other.take;
    }
}

export class Move {
    public movements: Movement[] = [];

    public static parse(json) {
        return new Move(json.movements.map(mv => Movement.parse(mv)));
    }

    constructor(movements?: Movement[]) {
        this.movements = movements || []
    }

    public equals(other: Move) {
        if (!other || other.movements.length !== this.movements.length) {
            return false;
        }

        return this.movements.every((mv, i) => mv.equals(other.movements[i]));
    }
}

export default Move;
